File: vcmrtd/example/ios/Runner/ProofingDeepLinkPlugin.swift
import Flutter
import UIKit

/**
 * Forwards a tapped vcmrtd:// identity-proofing deep link to Dart. Kept
 * separate from DeepLinkPlugin, which only accepts an https URL carrying
 * sessionId/nonce for the unrelated passport-issuer.yivi.app flow and would
 * reject this shape outright (see identity-proofing-service's deepLink() in
 * backend/internal/api/sessions.go — "vcmrtd://verify?token=...&api=...").
 */
@objc class ProofingDeepLinkPlugin: NSObject, FlutterPlugin {
    private static let channelName = "proofing_deeplink"

    private var channel: FlutterMethodChannel?
    private var initialLink: String?

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
        let instance = ProofingDeepLinkPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getInitialLink":
            result(initialLink)
            initialLink = nil // only ever answered once per cold start

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension ProofingDeepLinkPlugin {

    /// Cold start via a vcmrtd:// link: stash it for Dart's first
    /// getInitialLink call, since Flutter isn't ready to receive a method
    /// channel invocation yet at this point.
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        if let url = launchOptions?[.url] as? URL, url.scheme == "vcmrtd" {
            initialLink = url.absoluteString
        }
        return true
    }

    /// The app was already running when the link was tapped.
    func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
        guard url.scheme == "vcmrtd" else { return false }
        if let channel = channel {
            channel.invokeMethod("onLink", arguments: url.absoluteString)
        } else {
            initialLink = url.absoluteString
        }
        return true
    }
}

File: idem/android/app/src/main/kotlin/foundation/privacybydesign/idem/ProofingDeepLinkPlugin.kt
package foundation.privacybydesign.idem

import android.content.Intent
import androidx.annotation.NonNull
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Forwards a tapped vcmrtd:// identity-proofing deep link to Dart. Kept
 * separate from DeepLinkPlugin, which only accepts an https URL carrying
 * sessionId/nonce for the unrelated passport-issuer.yivi.app flow and would
 * reject this shape outright (see identity-proofing-service's deepLink() in
 * backend/internal/api/sessions.go — "vcmrtd://verify?token=...&api=...").
 */
class ProofingDeepLinkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    companion object {
        private const val CHANNEL = "proofing_deeplink"
    }

    private lateinit var channel: MethodChannel
    private var activity: FlutterActivity? = null
    private var initialLink: String? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as? FlutterActivity
        activity?.intent?.let { handleIntent(it) }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity as? FlutterActivity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getInitialLink" -> {
                result.success(initialLink)
                initialLink = null // only ever answered once per cold start
            }
            else -> result.notImplemented()
        }
    }

    /** The app was already running when the link was tapped. */
    fun onNewIntent(intent: Intent) {
        val url = extractUrl(intent) ?: return
        if (::channel.isInitialized) {
            channel.invokeMethod("onLink", url)
        } else {
            initialLink = url
        }
    }

    /** Cold start via a vcmrtd:// link. Reopening the app from recents
     *  replays the intent it was first launched with - an already used
     *  claim/handover link - so that one is ignored. */
    private fun handleIntent(intent: Intent) {
        if (intent.flags and Intent.FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY != 0) return
        initialLink = extractUrl(intent)
    }

    private fun extractUrl(intent: Intent?): String? {
        if (intent?.action != Intent.ACTION_VIEW) return null
        val uri = intent.data ?: return null
        if (uri.scheme != "vcmrtd") return null
        return uri.toString()
    }
}
